irohad/ametsuchi/impl/storage_impl.cpp

```diff
@@ -237,19 +253,30 @@ DROP TABLE IF EXISTS index_by_id_height_asset;
     expected::Result<std::shared_ptr<StorageImpl>, std::string>
     StorageImpl::create(std::string block_store_dir,
                         std::string postgres_options) {
-      auto ctx_result = initConnections(block_store_dir, postgres_options);
-      expected::Result<std::shared_ptr<StorageImpl>, std::string> storage;
-      ctx_result.match(
-          [&](expected::Value<ConnectionContext> &ctx) {
-            storage = expected::makeValue(std::shared_ptr<StorageImpl>(
-                new StorageImpl(block_store_dir,
-                                postgres_options,
-                                std::move(ctx.value.block_store),
-                                std::move(ctx.value.pg_lazy),
-                                std::move(ctx.value.pg_nontx))));
-          },
-          [&](expected::Error<std::string> &error) { storage = error; });
-      return storage;
+      return PostgresOptions::create(postgres_options) |
+                 [&block_store_dir](const PostgresOptions &options)
+                 -> expected::Result<std::shared_ptr<StorageImpl>,
+                                     std::string> {
+        options.getOption("dbname") | [&options](const auto &dbname) {
```

Comment:
On error case should be a error I guess, so use match

Reply:
Or maybe just use in a chain of results
get_dbname (bind)-> create db (and_res)-> initConnection (and_res)-> call ctor

Reply:
getOption returns optional. If it is empty, i.e. dbname param does not exist, we do not need to create the database

Reply:
Indeed, missed it